### ISSUE TITLE
Port old combinator implementations of sum types, iteration and recursion to new lambda-calculus style

### ISSUE THREAD

**Issue by VMatthijs** [VMatthijs/CHAD#4]
Also, some of the array primitives we had?

**Comment by VMatthijs** [comment]
(No rush, obviously)

**Comment by tomsmeding** [comment]
As for array primitives, we had only `map` and `foldr` before. Not that this is a reason to not have more, but just for posterity.

**Comment by VMatthijs** [comment]
Right! You did all the other ones (generate, permute, backpermute, indexing) in your own branch with dynamically sized arrays.

**Comment by VMatthijs** [comment]
@tomsmeding , want to try doing this at least for the sum types, so you can check whether it has the right complexity? That would then give us at least a way of doing AD on higher order programs with the right complexity by using defunctionalization.

**Comment by tomsmeding** [comment]
Oh that's a good one! I will do so. Slightly longer timeframe than usual perhaps due to lectures that I still need to make...

**Comment by tomsmeding** [comment]
@VMatthijs Seems to work with the classic LEither, but basically untested: https://github.com/VMatthijs/CHAD/compare/main...sum-types
Suggestions for good test cases?
